cluster/gce/util.sh

```diff
@@ -1066,6 +1066,18 @@ ETCD_APISERVER_SERVER_KEY: $(yaml-quote ${ETCD_APISERVER_SERVER_KEY_BASE64:-})
 ETCD_APISERVER_SERVER_CERT: $(yaml-quote ${ETCD_APISERVER_SERVER_CERT_BASE64:-})
 ETCD_APISERVER_CLIENT_KEY: $(yaml-quote ${ETCD_APISERVER_CLIENT_KEY_BASE64:-})
 ETCD_APISERVER_CLIENT_CERT: $(yaml-quote ${ETCD_APISERVER_CLIENT_CERT_BASE64:-})
+PROXY_SERVER_CA_KEY: $(yaml-quote ${PROXY_SERVER_CA_KEY_BASE64:-})
+PROXY_SERVER_CA_CERT: $(yaml-quote ${PROXY_SERVER_CA_CERT_BASE64:-})
+PROXY_SERVER_CERT: $(yaml-quote ${PROXY_SERVER_CERT_BASE64:-})
+PROXY_SERVER_KEY: $(yaml-quote ${PROXY_SERVER_KEY_BASE64:-})
+PROXY_SERVER_CLIENT_CERT: $(yaml-quote ${PROXY_SERVER_CLIENT_CERT_BASE64:-})
+PROXY_SERVER_CLIENT_KEY: $(yaml-quote ${PROXY_SERVER_CLIENT_KEY_BASE64:-})
+PROXY_AGENT_CA_KEY: $(yaml-quote ${PROXY_AGENT_CA_KEY_BASE64:-})
```

Comment:
Yeah I don't see why you can't stuff the agent pki into a secret?

Reply:
Or get rid of it. There's no reason the proxy can't use service account tokens to authenticate to the master. Unclear what the point of a separate CA is here.

Reply:
@mikedanese the connectivity goes the other way:

agent dials server.
apiserver dials server.
server proxies apiserver out through an agent.

Additionally we wish for the apiserver to be confident it's talking to a real server, and for the server to be confident it's receiving calls from a real apiserver. (same for the agent/server path).

I don't understand how tokens can serve this security model.

Reply:
Speaking for GCE deployment.

I'm mostly concerned with agent -> proxy server. My understanding is that the agent makes grpc requests to the server.
* For client authentication on this path, we can give the agent access to service account creds which can be forwarded in an authorization header. The proxy server can then perform a TokenReview and delegate authentication to the apiserver.
* For agent authenticating the proxy server, it depends on how much we trust the proxy server. Sharing the serving cert between the apiserver and proxy server is the easiest thing to do. If we don't trust the proxy server enough, we have a number of options to provision it a cert from the cluster root CA. The cluster root CA is good enough to get the kubelets to trust the apiserver, then it's probably good enough for this.

For the apiserver -> proxy server, do we expect these to be collocated for the foreseeable future? If so, we should use a unix domain socket. If that's good enough for apiserver to send all secrets to a KMS plugin, then that's good enough for this. If it's non-local:
* For client authentication, we can use the apiserver client cert (which grants the apiserver access to kubelet exec handlers among other things) or use service account tokens as proposed in https://github.com/kubernetes/enhancements/pull/658 .
* For apiserver authenticating the proxy server, we can do the same thing as path 1 bullet 2.

Every root of trust we've added (cluster root, etcd, aggregated API CA, service account credential issuer) has had different semantics. Each needs secure distribution, life cycle management (e.g. rotation of credentials chained up to the root, and rotation of the root itself) and needs to be made reliable. If we can reuse an existing root(s) in a way that meets our requirements, we can consolidate our efforts and improvements we make have broader impact.

Reply:
So I think we agree that we need to pick this up in the Beta Requirements for the KEP.

Reply:
Yup, beta is fine.